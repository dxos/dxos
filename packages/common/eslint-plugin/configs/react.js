//
// Copyright 2022 DXOS.org
//

module.exports = {
  extends: [
    'plugin:react/recommended',
    // TODO(burdon): Carefully test before enabling.
    // https://github.com/facebook/react/tree/main/packages/eslint-plugin-react-hooks
    // 'plugin:react-hooks/recommended',
  ],
  rules: {
    'jsx-quotes': ['error', 'prefer-single'],
    'react/display-name': 'off',
    'react/function-component-definition': [
      'error',
      {
        namedComponents: 'arrow-function',
        unnamedComponents: 'arrow-function',
      },
    ],
    'react/jsx-first-prop-new-line': ['error', 'multiline-multiprop'],
    'react/jsx-tag-spacing': [
      'error',
      {
        closingSlash: 'never',
        beforeSelfClosing: 'always',
        afterOpening: 'never',
        beforeClosing: 'never',
      },
    ],
    'react/jsx-wrap-multilines': 'off',
    'react/prop-types': 'off',
  },
  settings: {
    react: {
      version: 'detect',
    },
  },
  env: {
    browser: true,
  },
};
