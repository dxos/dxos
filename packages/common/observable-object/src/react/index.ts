//
// Copyright 2023 DXOS.org
//

export * from './observer';
export * from './useSubscription';
