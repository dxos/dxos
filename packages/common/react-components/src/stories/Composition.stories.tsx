//
// Copyright 2022 DXOS.org
//

import '@dxosTheme';
import { PaperPlane } from 'phosphor-react';
import React from 'react';

import { Button, DensityProvider, ElevationProvider, Input } from '../components';
import { defaultGroup, getSize } from '../styles';
import { mx } from '../util';

export default {
  component: DensityProvider
};

const ScenarioContent = (_props: {}) => (
  <>
    <Button>Text</Button>
    <Input
      label='excluded'
      variant='subdued'
      labelVisuallyHidden
      placeholder='Type something'
      defaultValue='Editable text'
    />
    <Button>
      <PaperPlane className={getSize(6)} />
    </Button>
  </>
);

const Scenario = (_props: {}) => (
  <>
    <div role='row' className={mx(defaultGroup({ elevation: 'group' }), 'flex gap-1 items-center p-0 mbe-4')}>
      <DensityProvider density='coarse'>
        <ElevationProvider elevation='chrome'>
          <ScenarioContent />
        </ElevationProvider>
      </DensityProvider>
    </div>
    <div role='row' className={mx(defaultGroup({ elevation: 'group' }), 'flex gap-1 items-center p-0')}>
      <DensityProvider density='fine'>
        <ElevationProvider elevation='chrome'>
          <ScenarioContent />
        </ElevationProvider>
      </DensityProvider>
    </div>
  </>
);

export const Default = {
  args: {},
  render: (_args: {}) => <Scenario />
};
