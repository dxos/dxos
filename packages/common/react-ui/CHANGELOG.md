# Changelog

## [0.1.0](https://github.com/dxos/dxos/compare/react-ui-v0.0.1...react-ui-v0.1.0) (2022-11-03)


### Features

* Add WRTC support to remote client ([#1656](https://github.com/dxos/dxos/issues/1656)) ([5666575](https://github.com/dxos/dxos/commit/5666575f35a09e6f476bb30c65ff031eb05d161d))
* Composer ([#1672](https://github.com/dxos/dxos/issues/1672)) ([d643cfe](https://github.com/dxos/dxos/commit/d643cfe3095ac206aa2d94736bef4d0b7b9e9ca3))
* Design system plugin, components, and assets ([#1596](https://github.com/dxos/dxos/issues/1596)) ([8a3acac](https://github.com/dxos/dxos/commit/8a3acacd15a86fce7a40dea5345a2eb75e0dfaf7))
* HALO pattern coverage ([#1711](https://github.com/dxos/dxos/issues/1711)) ([424f36b](https://github.com/dxos/dxos/commit/424f36bd37f02682e62fef56c87deaa3fc4298e8))
* Lists, invitations lifecycle, and other improvements ([#1777](https://github.com/dxos/dxos/issues/1777)) ([ce42706](https://github.com/dxos/dxos/commit/ce427069f71e29addfc42aab0a6d4327c7aa2048))
* NX Toolbox ([#1728](https://github.com/dxos/dxos/issues/1728)) ([bc1fd02](https://github.com/dxos/dxos/commit/bc1fd02c0e049576d2e7d0329f8ff50f4cfaefef))
* UI kit stub ([#1690](https://github.com/dxos/dxos/issues/1690)) ([ac4daca](https://github.com/dxos/dxos/commit/ac4daca50f87e3cfb5ab2aeebcda00046001c89b))
* Use react-uikit for HALO ([#1725](https://github.com/dxos/dxos/issues/1725)) ([251bebe](https://github.com/dxos/dxos/commit/251bebe7a0f6573be13b22d2ec8dee08e97e0029))


### Bug Fixes

* Logging in browser ([#1781](https://github.com/dxos/dxos/issues/1781)) ([b67a4c0](https://github.com/dxos/dxos/commit/b67a4c0b4e67b06bc4951c7a7000ee472f9d5b04))
* rename shims to typings ([#1730](https://github.com/dxos/dxos/issues/1730)) ([7e0868d](https://github.com/dxos/dxos/commit/7e0868d438c194b1fee03b34a8fadd3ab647d55c))
