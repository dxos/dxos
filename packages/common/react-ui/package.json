{
  "name": "@dxos/react-ui",
  "version": "0.1.13",
  "description": "Vite plugin which configures a low-level design system for DXOS.",
  "homepage": "https://dxos.org",
  "bugs": "https://github.com/dxos/dxos/issues",
  "license": "MIT",
  "author": "DXOS.org",
  "main": "dist/lib/browser/index.mjs",
  "types": "dist/types/src/index.d.ts",
  "files": [
    "dist",
    "src",
    "plugin.js",
    "plugin.d.ts"
  ],
  "dependencies": {
    "@fontsource/fira-code": "^4.5.12",
    "@fontsource/roboto-flex": "^4.5.2",
    "@fontsource/space-grotesk": "^4.5.10",
    "@headlessui/react": "^1.7.3",
    "@radix-ui/react-alert-dialog": "^1.0.2",
    "@radix-ui/react-avatar": "^1.0.1",
    "@radix-ui/react-dialog": "^1.0.0",
    "@radix-ui/react-navigation-menu": "^1.1.1",
    "@radix-ui/react-popover": "^1.0.2",
    "@radix-ui/react-portal": "^1.0.1",
    "@radix-ui/react-toast": "^1.1.1",
    "@radix-ui/react-toolbar": "^1.0.1",
    "@radix-ui/react-tooltip": "^1.0.2",
    "@tailwindcss/forms": "^0.5.3",
    "alea": "^1.0.1",
    "autoprefixer": "^10.4.12",
    "jdenticon": "^3.2.0",
    "postcss": "^8.4.17",
    "qrcode.react": "^3.1.0",
    "rci": "^0.0.3",
    "tailwind-merge": "^1.8.0",
    "tailwindcss": "^3.1.8",
    "tailwindcss-logical": "^3.0.0",
    "tailwindcss-radix": "^2.6.0"
  },
  "devDependencies": {
    "@types/react": "^18.0.21",
    "@types/react-dom": "^18.0.6",
    "@vitejs/plugin-react": "^2.0.1",
    "esbuild": "~0.14.1",
    "esbuild-style-plugin": "^1.6.0",
    "phosphor-react": "^1.4.1",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "storybook-dark-mode": "^1.1.2",
    "vite": "3.0.9"
  },
  "peerDependencies": {
    "phosphor-react": "^1.4.1",
    "react": "^18.2.0",
    "react-dom": "^18.2.0"
  },
  "publishConfig": {
    "access": "public"
  }
}
