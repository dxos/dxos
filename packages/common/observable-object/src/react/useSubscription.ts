//
// Copyright 2022 DXOS.org
//

import { useEffect, useRef } from 'react';

import { Selection, SubscriptionHandle, createSubscription } from '../access-observer';

/**
 * Create reactive selection.
 * Calls the callback when the selection changes and during the first render.
 */
export const useSubscription = (cb: () => void, selection: Selection) => {
  // Make sure that the callback is always the one from the latest render.
  // Without this, we would always call the callback from the initial render,
  // which might contain stale data.
  const callbackRef = useRef(cb);
  callbackRef.current = cb;

  const subscriptionRef = useRef<SubscriptionHandle>();

  useEffect(() => {
    subscriptionRef.current = createSubscription(() => {
      callbackRef.current();
    });

    return () => subscriptionRef.current?.unsubscribe();
  }, []);

  subscriptionRef.current?.update(selection);
  return subscriptionRef.current;
};
