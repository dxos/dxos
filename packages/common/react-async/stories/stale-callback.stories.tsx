//
// Copyright 2021 DXOS.org
//

import React, { useEffect, useState } from 'react';

import { useStateWithRef } from '../src';

export default {
  title: 'react-async/stale-callback'
};

export const Primary = () => {
  const [value, setValue, valueRef] = useStateWithRef(0);
  const [result, setResult] = useState({});

  useEffect(() => {
    let counter = 0;
    const i = setInterval(() => {
      setValue(++counter);
    }, 1000);

    return () => clearInterval(i);
  }, []);

  useEffect(() => {
    const t = setTimeout(() => {
      setResult({
        value, // Value will be 0.
        ref: valueRef.current // Ref will be 5.
      });
    }, 5000);

    return () => clearTimeout(t);
  }, []);

  return (
    <div style={{ display: 'flex', alignItems: 'center', padding: 16 }}>
      <div style={{ paddingLeft: 16 }}>Value: {value}</div>
      <div style={{ paddingLeft: 16 }}>Result: {JSON.stringify(result)}</div>
    </div>
  );
};
