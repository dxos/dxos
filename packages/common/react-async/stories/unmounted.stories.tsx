//
// Copyright 2021 DXOS.org
//

import React, { useEffect, useState } from 'react';

import { useMounted } from '../src';

export default {
  title: 'react-async/unmounted'
};

const AsyncComponent = ({ nocheck = false }: { nocheck?: boolean }) => {
  const [value, setValue] = useState('Processing...');
  const isMounted = useMounted();
  useEffect(() => {
    setTimeout(() => {
      if (nocheck || isMounted()) {
        // Check if still mounted.
        setValue('Done');
      }
    }, 2000);
  }, []);

  return <div>{value}</div>;
};

const TestApp = () => {
  const [show, setShow] = useState(true);
  useEffect(() => {
    setTimeout(() => {
      setShow(false); // Remove element.
    }, 1000);
  }, []);

  if (!show) {
    return null;
  }

  return <AsyncComponent nocheck />;
};

export const Primary = () => (
  <div style={{ padding: 16 }}>
    <TestApp />
  </div>
);
