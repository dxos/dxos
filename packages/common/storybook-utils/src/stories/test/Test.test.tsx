//
// Copyright 2025 DXOS.org
//

import { composeStories } from '@storybook/react';
import { fireEvent, render, screen } from '@testing-library/react';
import React from 'react';
import { describe, expect, it, vi } from 'vitest';

import { ThemeProvider } from '@dxos/react-ui';

import { TEST_ID, Test } from './Test';
import * as stories from './Test.stories';

const { Default } = composeStories(stories);

/**
 * Vitest sanity test (should be visible in the storybook).
 * https://storybook.js.org/docs/writing-tests/integrations/vitest-addon
 */
describe('Test', () => {
  it('should render', async () => {
    await Default.run();
    expect(screen.getByTestId(TEST_ID)).to.exist;
  });

  it('calls onClick when clicked', () => {
    const handleClick = vi.fn();
    const { container } = render(
      <ThemeProvider>
        <Test id='test' icon='ph--x--regular' label='Test' onClick={handleClick} />
      </ThemeProvider>,
    );

    fireEvent.click(container.querySelector('#test')!);
    expect(handleClick).toHaveBeenCalledTimes(1);
    expect(screen.debug()).toMatchSnapshot();
  });
});
