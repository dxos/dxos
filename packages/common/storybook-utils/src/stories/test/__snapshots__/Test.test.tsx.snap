// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test > calls onClick when clicked 1`] = `undefined`;
